Allow/forbid certain block interaction from the config.

Raycoms opened the pull request:
Closes #686 
Closes #667 


# Changes proposed in this pull request:
- Classic firewall rule, block everything which isn't defined otherwise.
- Allows interaction with certain blocks defined in the config.
- Adds a gui to edit permission of ranks
- Adds a gui to add blocks or positions.
- Rework the event handling again.

Review please

Raycoms committed:
Allow/forbid certain block interaction from the config.

MinecoloniesSonar commented:
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/permissions/ColonyPermissionEventHandler.java:
```diff
@@ -189,8 +190,14 @@ public void on(final PlayerInteractEvent event)
+                return;
+            }
+
             if(Configurations.enableColonyProtection && (event.getWorld().getBlockState(event.getPos()).getBlock() instanceof BlockContainer
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Add parentheses to make the operator precedence explicit. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS864)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Reduce the number of conditional operators (5) used in the expression (maximum allowed 4). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1067)

marchermans commented on src/main/java/com/minecolonies/coremod/configuration/ConfigurationHandler.java:
```diff
@@ -114,6 +114,12 @@ public static synchronized void init(final File file)
+            freeToInteractBlocks = config.get(CATEGORY_NAMES, "freeToInteractBlocks", freeToInteractBlocks,
+                    "Blocks players should be able to interact with inside any colony.").getStringList();
+
+            forbigInteractionBlocks = config.get(CATEGORY_NAMES, "forbigInteractionBlocks", forbigInteractionBlocks,
```
Do you not mean "forbidInteractionBlocks" ???

Raycoms committed:
refactor wrong type

Raycoms committed:
first adaptions to config

Raycoms committed:
Setup new permission properties and handle them in event handler.

Raycoms committed:
Fix up permission of other actions

- Scan tool
- Remove player from list
- Demote/promote 

To be able to make permission changes you must be of a higher rank than the rank you're changing. (If friends are allowed to change permission they can add hostiles but not remove other friends)

Raycoms committed:
Fix a bug and set up first views for the GUI

MinecoloniesSonar commented:
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/items/ItemScanTool.java:
```diff
@@ -1,6 +1,7 @@
 package com.minecolonies.coremod.items;
 
 import com.minecolonies.coremod.MineColonies;
+import com.minecolonies.coremod.colony.ColonyManager;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Remove this unused import 'com.minecolonies.coremod.colony.ColonyManager'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

Raycoms committed:
get the gui running

MinecoloniesSonar commented:
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java:
```diff
@@ -259,6 +344,16 @@
+    /**
+     * The ScrollingList of the users.
+     */
+    private ScrollingList freeBlocksList;
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove the "freeBlocksList" field and declare it as a local variable in the relevant methods. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1450)
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Perhaps 'freeBlocksList' could be replaced by a local variable. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ASingularField)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java:
```diff
@@ -290,9 +385,103 @@ public WindowTownHall(final BuildingTownHall.View townHall)
+
+    }
+
+    public void trigger(Button button)
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Document this public method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

Raycoms committed:
Implment the list of permissions and add language string support

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java:
```diff
@@ -475,6 +695,65 @@ public void updateElement(final int index, @NotNull final Pane rowPane)
+            @Override
+            public int getElementCount()
+            {
+                return Permissions.Action.values().length-3;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Assign this magic number 3 to a well-named constant, and use the constant instead. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java:
```diff
@@ -11,15 +11,19 @@
 import com.minecolonies.coremod.network.messages.ToggleJobMessage;
 import com.minecolonies.coremod.network.messages.WorkOrderChangeMessage;
 import com.minecolonies.coremod.util.LanguageHandler;
+import net.minecraft.entity.player.EntityPlayer;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Remove this unused import 'net.minecraft.entity.player.EntityPlayer'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java:
```diff
@@ -475,6 +695,65 @@ public void updateElement(final int index, @NotNull final Pane rowPane)
+                rowPane.findPaneOfTypeByID("trigger", Button.class)
+                        .setLabel(isTriggered ? LanguageHandler.format("com.minecolonies.coremod.gui.workerHuts.retrieveOn")
+                                : LanguageHandler.format("com.minecolonies.coremod.gui.workerHuts.retrieveOff"));
+                rowPane.findPaneOfTypeByID("index", Label.class).setLabelText(actionIndex + "");
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Use "Integer.toString" instead. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS2131)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java:
```diff
@@ -290,9 +395,123 @@ public WindowTownHall(final BuildingTownHall.View townHall)
+        {
+            index = Integer.valueOf(((Label) pane).getLabelText());
+        }
+        final boolean trigger = !(LanguageHandler.format("com.minecolonies.coremod.gui.workerHuts.retrieveOn")).equals(button.getLabel());
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Define a constant instead of duplicating this literal "com.minecolonies.coremod.gui.workerHuts.retrieveOn" 3 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1192)

Raycoms committed:
First part of permission handling working. Optimize imports

MinecoloniesSonar commented:
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

Raycoms committed:
Part 1 of populating the freeBlock and freePosition lists

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/Colony.java:
```diff
@@ -165,6 +178,17 @@ protected Colony(final int id, final int dim)
+
+        //todo transform things from configuration to list.
+        //todo need some ways to make it blockState really.
+        for(String s: Configurations.freeToInteractBlocks)
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Local variable 's' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java:
```diff
@@ -11,12 +11,13 @@
+import com.minecolonies.coremod.network.messages.*;
 import com.minecolonies.coremod.util.LanguageHandler;
+import net.minecraft.block.Block;
+import net.minecraft.block.state.IBlockState;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Remove this unused import 'net.minecraft.block.state.IBlockState'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java:
```diff
@@ -290,9 +412,167 @@ public WindowTownHall(final BuildingTownHall.View townHall)
+
+        final String[] strings = inputText.split(" ");
+
+        if(strings.length == 3)
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Assign this magic number 3 to a well-named constant, and use the constant instead. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)

MinecoloniesSonar commented:
SonarQube analysis reported 7 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 5 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Colony.java#L45](https://github.com/Minecolonies/minecolonies/blob/38f1e409321c6477b42ab162250ec7f3ce4220c8/src/main/java/com/minecolonies/coremod/colony/Colony.java#L45): class "Colony" has 53 public methods, which is greater than the 50 authorized. Split it into smaller classes. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1448)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

Raycoms committed:
Finish adding and removing.

- Missing testing and small adaptions.

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/network/messages/ChangeFreeToInteractBlockMessage.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    @Override
+    public void messageOnServerThread(final ChangeFreeToInteractBlockMessage message, final EntityPlayerMP player)
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') The Cyclomatic Complexity of this method "messageOnServerThread" is 12 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/Colony.java:
```diff
@@ -165,6 +179,15 @@ protected Colony(final int id, final int dim)
         // Register a new event handler
         MinecraftForge.EVENT_BUS.register(new ColonyPermissionEventHandler(this));
+
+        for(String s: Configurations.freeToInteractBlocks)
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Local variable 's' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

MinecoloniesSonar commented:
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Colony.java#L46](https://github.com/Minecolonies/minecolonies/blob/6c64784e7997c53490ddee3cee77ae605d8617d8/src/main/java/com/minecolonies/coremod/colony/Colony.java#L46): class "Colony" has 53 public methods, which is greater than the 50 authorized. Split it into smaller classes. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1448)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

Raycoms committed:
fix things up!

MinecoloniesSonar commented:
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Colony.java#L46](https://github.com/Minecolonies/minecolonies/blob/57f7080620db0ba9f117a846797ffb39dc8453de/src/main/java/com/minecolonies/coremod/colony/Colony.java#L46): class "Colony" has 53 public methods, which is greater than the 50 authorized. Split it into smaller classes. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1448)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

Raycoms committed:
sonar fixes

MinecoloniesSonar commented:
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Colony.java#L46](https://github.com/Minecolonies/minecolonies/blob/52c789164c0545105864fbccc369185b6a6f5c42/src/main/java/com/minecolonies/coremod/colony/Colony.java#L46): class "Colony" has 53 public methods, which is greater than the 50 authorized. Split it into smaller classes. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1448)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

marchermans commented on src/main/java/com/minecolonies/coremod/permissions/ColonyPermissionEventHandler.java:
```diff
@@ -262,9 +273,32 @@ public void on(final PlayerInteractEvent.EntityInteract event)
     @SubscribeEvent
     public void on(final PlayerInteractEvent.EntityInteractSpecific event)
     {
+        if(isFreeToInteractWith(event.getWorld().getBlockState(event.getPos()).getBlock()))
```
Why? Read the EntityInteractSpecific Javadoc:
/**
     * This event is fired on both sides whenever a player right clicks an **entity**.
     *
     * "Interact at" is an interact where the local vector (which part of the entity you clicked) is known.
     * The state of this event affects whether {@link Entity#applyPlayerInteraction} is called.
     * If {@link Entity#applyPlayerInteraction} returns {@link net.minecraft.util.EnumActionResult#SUCCESS}, then processing ends.
     * Otherwise processing will continue to {@link EntityInteract}
     *
     * Canceling the event clientside will cause processing to continue to {@link EntityInteract},
     * while canceling serverside will simply do no further processing.
     */
    @Cancelable
    public static class EntityInteractSpecific extends PlayerInteractEvent

It is for player with entity interaction. Block checking should not be done here!

marchermans commented on src/main/java/com/minecolonies/coremod/colony/ColonyView.java:
```diff
@@ -86,10 +97,80 @@ public static void serializeNetworkData(@NotNull final Colony colony, @NotNull f
+     */
+    public void addFreePosition(@NotNull final BlockPos pos)
+    {
+        freePositions.add(pos);
```
This should not happen on the client side.
Packet should be send to the server. Update of the view!

Raycoms replied:
I know the client will receive the message. But the view should also change instantly.
So if you don't want to wait until we the messages got sent around completely again. We make the changes on client and server side so that the user sees the change directly.

marchermans replied:
Understood. Accepted.

marchermans commented on src/main/java/com/minecolonies/coremod/colony/ColonyView.java:
```diff
@@ -86,10 +97,80 @@ public static void serializeNetworkData(@NotNull final Colony colony, @NotNull f
+     */
+    public void addFreeBlock(@NotNull final Block block)
+    {
+        freeBlocks.add(block);
```
Again. Server side interaction.

Raycoms replied:
Same here

marchermans commented on src/main/java/com/minecolonies/coremod/colony/ColonyView.java:
```diff
@@ -86,10 +97,80 @@ public static void serializeNetworkData(@NotNull final Colony colony, @NotNull f
+     */
+    public void removeFreePosition(@NotNull final BlockPos pos)
+    {
+        freePositions.remove(pos);
```
See above.

marchermans commented on src/main/java/com/minecolonies/coremod/colony/ColonyView.java:
```diff
@@ -86,10 +97,80 @@ public static void serializeNetworkData(@NotNull final Colony colony, @NotNull f
+     */
+    public void removeFreeBlock(@NotNull final Block block)
+    {
+        freeBlocks.remove(block);
```
See above.

marchermans commented on src/main/java/com/minecolonies/coremod/configuration/Configurations.java:
```diff
@@ -53,6 +53,12 @@
+    public static String[] freeToInteractBlocks = new String[]
+            {
+                    "minecraft:BlockDirt",
+                    "minecraft:damnIt"
```
Never heard of this block before.

Raycoms replied:
oh I completely forgot I had this test blocks in here.

marchermans commented on src/main/java/com/minecolonies/coremod/permissions/ColonyPermissionEventHandler.java:
```diff
@@ -186,27 +203,46 @@ public void on(final PlayerInteractEvent event)
-                    || event.getWorld().getTileEntity(event.getPos()) != null || (event.getItemStack() != null && event.getItemStack().getItem() instanceof ItemPotion)))
+            final Permissions perms = colony.getPermissions();
+
+            if(isFreeToInteractWith(event.getWorld().getBlockState(event.getPos()).getBlock(), event.getPos())
```
This method catches all PlayerInteractEvents. That means that it catches also the entity interact event. Handling this here is not a good idea.

marchermans commented on src/main/java/com/minecolonies/coremod/permissions/ColonyPermissionEventHandler.java:
```diff
@@ -246,7 +279,13 @@ private void checkEventCancelation(final Permissions.Rank rankIn, @NotNull final
     public void on(final PlayerInteractEvent.EntityInteract event)
     {
-        checkEventCancelation(Permissions.Rank.FRIEND, event.getEntityPlayer(), event.getWorld(), event);
+        if(isFreeToInteractWith(event.getWorld().getBlockState(event.getPos()).getBlock(), event.getPos())
```
Again for Entities. Block checking should not be done here!

marchermans commented on src/main/java/com/minecolonies/coremod/permissions/ColonyPermissionEventHandler.java:
```diff
@@ -262,9 +301,23 @@ public void on(final PlayerInteractEvent.EntityInteract event)
     public void on(final PlayerInteractEvent.EntityInteractSpecific event)
     {
-        checkEventCancelation(Permissions.Rank.FRIEND, event.getEntityPlayer(), event.getWorld(), event);
+        if(isFreeToInteractWith(event.getWorld().getBlockState(event.getPos()).getBlock(), event.getPos())
```
See above. Entity interaction.

marchermans reviewed (changes requested):
This will not work. You are also checking the block list when interacting with normal entities. Which will not work.

Raycoms committed:
Improve the handling with entity interaction.

MinecoloniesSonar commented:
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Colony.java#L46](https://github.com/Minecolonies/minecolonies/blob/aa62a639e42bd8bb69b378282823c547e750fac0/src/main/java/com/minecolonies/coremod/colony/Colony.java#L46): class "Colony" has 53 public methods, which is greater than the 50 authorized. Split it into smaller classes. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1448)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

marchermans reviewed (approved):
Approced

marchermans commented:
Damn Type!

Raycoms committed:
remove strange config blocks.

MinecoloniesSonar commented:
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Colony.java#L46](https://github.com/Minecolonies/minecolonies/blob/5da2c8a475979b75e97e60407452db7ec5ec7dd8/src/main/java/com/minecolonies/coremod/colony/Colony.java#L46): class "Colony" has 53 public methods, which is greater than the 50 authorized. Split it into smaller classes. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1448)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

Raycoms committed:
Improve handling

MinecoloniesSonar commented:
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Colony.java#L46](https://github.com/Minecolonies/minecolonies/blob/79462fe7a84d6a6a6a599dc057d201cda02e538e/src/main/java/com/minecolonies/coremod/colony/Colony.java#L46): class "Colony" has 53 public methods, which is greater than the 50 authorized. Split it into smaller classes. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1448)
1. ![MINOR][MINOR] [AbstractEntityAIStructure.java#L6](https://github.com/Minecolonies/minecolonies/blob/79462fe7a84d6a6a6a599dc057d201cda02e538e/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L6): Remove this unused import 'com.minecolonies.coremod.colony.buildings.AbstractBuildingWorker'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)
1. ![MINOR][MINOR] [AbstractEntityAIStructure.java#L9](https://github.com/Minecolonies/minecolonies/blob/79462fe7a84d6a6a6a599dc057d201cda02e538e/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L9): Remove this unused import 'com.minecolonies.coremod.colony.buildings.utils.BuildingBuilderResource'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/Colony.java:
```diff
@@ -165,6 +179,23 @@ protected Colony(final int id, final int dim)
+        for(final String s: Configurations.freeToInteractBlocks)
+        {
+            final Block block = Block.getBlockFromName(s);
+            if(block != null)
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Avoid if (x != y) ..; else ..; [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AConfusingTernary)

Raycoms committed:
optimize imports

Raycoms committed:
optimize a sonar issue

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/Colony.java:
```diff
@@ -165,6 +179,23 @@ protected Colony(final int id, final int dim)
+        for(final String s: Configurations.freeToInteractBlocks)
+        {
+            final Block block = Block.getBlockFromName(s);
+            if(block == null)
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Avoid if (x != y) ..; else ..; [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AConfusingTernary)

MinecoloniesSonar commented:
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Colony.java#L46](https://github.com/Minecolonies/minecolonies/blob/282c82c9785c84c3ecca21a487a8d3a31df9afdc/src/main/java/com/minecolonies/coremod/colony/Colony.java#L46): class "Colony" has 53 public methods, which is greater than the 50 authorized. Split it into smaller classes. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1448)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

Raycoms committed:
Fix a nullpointer and add a bit more permission editing

Player should be allowed to demote or remove himself.

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/network/messages/PermissionsMessage.java:
```diff
@@ -369,8 +374,11 @@ public void messageOnServerThread(final RemovePlayer message, final EntityPlayer
             final Permissions.Player permissionsPlayer = colony.getPermissions().getPlayers().get(message.playerID);
-            if ((permissionsPlayer.getRank() == Permissions.Rank.HOSTILE && colony.getPermissions().hasPermission(player, Permissions.Action.CAN_PROMOTE))
-                  || (permissionsPlayer.getRank() != Permissions.Rank.HOSTILE && colony.getPermissions().hasPermission(player, Permissions.Action.CAN_DEMOTE)))
+            if ((permissionsPlayer.getRank() == Permissions.Rank.HOSTILE && colony.getPermissions().hasPermission(player, Permissions.Action.EDIT_PERMISSIONS))
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Reduce the number of conditional operators (5) used in the expression (maximum allowed 4). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1067)

MinecoloniesSonar commented:
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Colony.java#L46](https://github.com/Minecolonies/minecolonies/blob/a97d7ee13ac56dd34babc787fbe246efe2db05d3/src/main/java/com/minecolonies/coremod/colony/Colony.java#L46): class "Colony" has 53 public methods, which is greater than the 50 authorized. Split it into smaller classes. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1448)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

marvin-bitterlich commented on src/main/java/com/minecolonies/coremod/configuration/ConfigurationHandler.java:
```diff
@@ -114,6 +114,9 @@ public static synchronized void init(final File file)
                     "Last Names").getStringList();
+
+            freeToInteractBlocks = config.get(CATEGORY_NAMES, "freeToInteractBlocks", freeToInteractBlocks,
+                    "Blocks players should be able to interact with inside any colony.").getStringList();
```
> Block and Positions
should be added to avoid confusion with the 0 0 0

marvin-bitterlich commented on src/main/java/com/minecolonies/coremod/network/messages/ChangeFreeToInteractBlockMessage.java:
```diff
@@ -0,0 +1,153 @@
+
+            if(message.type == MessageType.ADD_BLOCK)
+            {
+                if (message.block != Blocks.DIRT)
```
change this to only use 0 0 0 as null value, not both

marvin-bitterlich replied:
On second thought we should use multiple messages

marvin-bitterlich commented on src/main/java/com/minecolonies/coremod/permissions/ColonyPermissionEventHandler.java:
```diff
@@ -97,12 +106,18 @@ private boolean checkBlockEventDenied(final World worldIn, final BlockPos posIn,
     }
 
-    private static void cancelEvent(final Event event)
+    private static void cancelEvent(final Event event, @Nullable final EntityPlayer player)
```
if you have the event, maybe you can get the player from it inside this method?

Raycoms replied:
No we can't because every event gets the player differently.
So we can't actually get the player from the plain event.

marvin-bitterlich replied:
kk

marvin-bitterlich commented on src/main/java/com/minecolonies/coremod/permissions/ColonyPermissionEventHandler.java:
```diff
@@ -161,7 +178,7 @@ public void on(final ExplosionEvent.Start event)
                 && colony.isCoordInColony(event.getWorld(), new BlockPos(event.getExplosion().getPosition())))
         {
-            cancelEvent(event);
+            cancelEvent(event, null);
```
explosion start should alert the player if he caused it

Raycoms replied:
no way to get the player object from the explosion event.

marvin-bitterlich reviewed (changes requested):
reviewed all but gui.

Raycoms committed:
improve the message for freeBlocks

Raycoms committed:
Merge remote-tracking branch 'refs/remotes/origin/develop'

Conflicts:
	src/main/java/com/minecolonies/coremod/util/BlockPosUtil.java

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/network/messages/ChangeFreeToInteractBlockMessage.java:
```diff
@@ -0,0 +1,150 @@
+
+            if(message.type == MessageType.ADD_BLOCK)
+            {
+                if (!(message.pos.getX() == 0 && message.pos.getZ() == 0 && message.pos.getY() == 0))
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Avoid if (x != y) ..; else ..; [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AConfusingTernary)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/network/messages/ChangeFreeToInteractBlockMessage.java:
```diff
@@ -0,0 +1,150 @@
+import net.minecraftforge.fml.common.network.ByteBufUtils;
+import net.minecraftforge.fml.common.network.simpleimpl.IMessage;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Remove this unused import 'org.jetbrains.annotations.Nullable'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/network/messages/ChangeFreeToInteractBlockMessage.java:
```diff
@@ -0,0 +1,150 @@
+            }
+            else
+            {
+                if (!(message.pos.getX() == 0 && message.pos.getZ() == 0 && message.pos.getY() == 0))
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Avoid if (x != y) ..; else ..; [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AConfusingTernary)

MinecoloniesSonar commented:
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Colony.java#L46](https://github.com/Minecolonies/minecolonies/blob/3d2d8ac5f3a9b7f1a4b81301de8b573d10b29be8/src/main/java/com/minecolonies/coremod/colony/Colony.java#L46): class "Colony" has 53 public methods, which is greater than the 50 authorized. Split it into smaller classes. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1448)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

Raycoms committed:
remove unused import

MinecoloniesSonar commented:
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Colony.java#L46](https://github.com/Minecolonies/minecolonies/blob/c82a12a0d44552a48a89fb4296da250a5d2c5591/src/main/java/com/minecolonies/coremod/colony/Colony.java#L46): class "Colony" has 53 public methods, which is greater than the 50 authorized. Split it into smaller classes. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1448)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

marvin-bitterlich commented:
LGTM

[![Approved with PullApprove](https://img.shields.io/badge/maintainers-approved-brightgreen.svg)](https://pullapprove.com/Minecolonies/minecolonies/pull-request/690/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)
